### PR TITLE
Switch to installing/uninstalling nuget packages in the BG instead of the UI thread

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes [AB#959448](https://devdiv.visualstudio.com/0bdbc590-a062-4c3f-b0f6-9383f67865ee/_workitems/edit/959448)
Blocked on https://github.com/NuGet/Home/issues/11021 where Nuget documents that what we are doing is explicitly ok.

### CODE REVIEW

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs [hunk 1]
```diff
@@ -482,6 +512,8 @@ private void OnWorkspaceChanged(object sender, WorkspaceChangeEventArgs e)
         private async Task<T?> PerformNuGetProjectServiceWorkAsync<T>(
             Func<INuGetProjectService, CancellationToken, ValueTask<T?>> doWorkAsync, CancellationToken cancellationToken)
         {
+            await ThreadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs:515]: i don't know if this is necessary.   but i was trying to preserve prior semantics.  @AArnott do you need to be on the UI thread to call .GetServiceAsync to get the IBrokeredServiceContainer?

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs [hunk 2]
```diff
@@ -306,28 +317,32 @@ private void OnSourceProviderSourcesChanged(object sender, EventArgs e)
...
+            EnvDTE.Project dteProject,
+            IProgressTracker progressTracker,
+            CancellationToken cancellationToken)
         {
             this.AssertIsForeground();
             Contract.ThrowIfFalse(IsEnabled);
 
-            // TODO: consider putting this under a TWD so it would actually be cancellable.
-            var cancellationToken = CancellationToken.None;
+            var description = string.Format(ServicesVSResources.Installing_0, packageName);
```
Comment (AArnott) [inline src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs:325]: An analyzer exists that would flag this line because you don't provide a `CultureInfo` object as a first argument. I recommend it even if you aren't running the analyzer. Either `CultureInfo.InvariantCulture` (for machine consumption) or `CultureInfo.CurrentCulture` (for human consumption).

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs [hunk 3]
```diff
@@ -306,28 +317,32 @@ private void OnSourceProviderSourcesChanged(object sender, EventArgs e)
...
+            progressTracker.Description = description;
+            dte.StatusBar.Text = description;
 
             try
             {
-                return this.ThreadingContext.JoinableTaskFactory.Run(() =>
-                {
-                    return this.PerformNuGetProjectServiceWorkAsync(async (nugetService, cancellationToken) =>
+                return this.ThreadingContext.JoinableTaskFactory.Run(
```
Comment (AArnott) [inline src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs:331]: Is now a good time to consider whether blocking the caller during this async work is really necessary?
You remove a SwitchToMainThreadAsync and add a switch to the threadpool, but if this code runs while a JTF.Run method is blocking our caller (who we verify above *is* on the main thread), then it's all moot because you'll block the main thread during the whole thing anyway.

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs [hunk 4]
```diff
@@ -365,35 +386,44 @@ private static string GetStatusBarText(string packageName, string? installedVers
...
+                    () => this.PerformNuGetProjectServiceWorkAsync(async (nugetService, cancellationToken) =>
                     {
-                        await this.ThreadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
```
Comment (AArnott) [inline src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs:382]: Same feedback as above: aren't you still blocking the main thread by virtue of being inside a JTF.Run that we know was called from the main thread?

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs (hunk 1 above)
Comment (AArnott) [reply src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs:513]: No. GetServiceAsync can be called from a background thread. But it's generally *not* safe to cast the result as you're doing unless you are on the main thread or know that the service object is an MTA object. In this case, it *is* an MTA so you're fine.

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs [hunk 5]
```diff
@@ -482,6 +512,8 @@ private void OnWorkspaceChanged(object sender, WorkspaceChangeEventArgs e)
         private async Task<T?> PerformNuGetProjectServiceWorkAsync<T>(
             Func<INuGetProjectService, CancellationToken, ValueTask<T?>> doWorkAsync, CancellationToken cancellationToken)
         {
+            await ThreadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+
             var serviceContainer = (IBrokeredServiceContainer?)await _asyncServiceProvider.GetServiceAsync(typeof(SVsBrokeredServiceContainer)).ConfigureAwait(false);
```
Comment (AArnott) [inline src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs:515]: But because of that caveat, we recommend folks switch to using [the extension method](https://docs.microsoft.com/en-us/dotnet/api/microsoft.visualstudio.shell.serviceextensions.getserviceasync?view=visualstudiosdk-2019#Microsoft_VisualStudio_Shell_ServiceExtensions_GetServiceAsync__2_Microsoft_VisualStudio_Shell_IAsyncServiceProvider_System_Boolean_):
```suggestion
            IBrokeredServiceContainer? serviceContainer = await _asyncServiceProvider.GetServiceAsync<SVsBrokeredServiceContainer, IBrokeredServiceContainer>(throwOnFailure: false).ConfigureAwait(false);
```

**Review by AArnott — APPROVED** [review]

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs:325]: i don't think we ever provide those for loc strings, so i'm going to consider that out of scope for this change.

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs (hunk 3 above)
Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs:331]: the main issue here is that this is called from the undo/redo stack.  and it hurts my brain to try to consider what it means to do undo/redo asynchronously.

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs (hunk 5 above)
Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs:515]: > No. GetServiceAsync can be called from a background thread. But it's generally not safe to cast the result as you're doing unless you are on the main thread or know that the service object is an MTA object. In this case, it is an MTA so you're fine.

Ah ok.

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs (hunk 4 above)
Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs:382]: yeah, i will fix this all up :)

**Review by AArnott — APPROVED** [review]

File: src/Features/Core/Portable/AddPackage/InstallPackageDirectlyCodeActionOperation.cs [hunk 6]
```diff
@@ -72,7 +72,8 @@ internal class InstallPackageDirectlyCodeActionOperation : CodeActionOperation
         {
             return _installerService.TryInstallPackage(
                 workspace, _document.Id, _source, _packageName,
-                _versionOpt, _includePrerelease, cancellationToken);
+                _versionOpt, _includePrerelease,
+                progressTracker, cancellationToken);
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/AddPackage/InstallPackageDirectlyCodeActionOperation.cs:76]: once https://github.com/dotnet/roslyn/pull/55682 goes in, this can be async end to end.

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory_UndoRedo.cs [hunk 7]
```diff
@@ -72,17 +81,37 @@ private abstract class BaseUndoUnit : IOleUndoUnit
...
             }
 
-            public abstract void Do(IOleUndoManager pUndoManager);
+            protected abstract Task DoWorkerAsync(IOleUndoManager pUndoManager);
             public abstract void GetDescription(out string pBstr);
 
+            public void Do(IOleUndoManager pUndoManager)
+            {
+                // Fire and forget
+                _ = DoAsync(pUndoManager);
```
Comment (sharwell) [inline src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory_UndoRedo.cs:98]: ❔ Missing async token?

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs [hunk 8]
```diff
@@ -487,9 +530,6 @@ private void OnWorkspaceChanged(object sender, WorkspaceChangeEventArgs e)
             if (serviceBroker == null)
                 return default;
 
-            // Make sure we are on the thread pool to avoid UI thread dependencies if external code uses ConfigureAwait(true)
-            await TaskScheduler.Default;
```
Comment (sharwell) [inline src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs:491]: ❓ Why was this line removed?

**Review by sharwell — CHANGES_REQUESTED** [review]
Had a question

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs (hunk 8 above)
Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs:491]: i'll check again, but it should be happenign higher up (and everything is CA(false) after that), so it was redundant.

Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs:491]: added back with comment.

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs [hunk 9]
```diff
@@ -361,42 +391,56 @@ private void OnSourceProviderSourcesChanged(object sender, EventArgs e)
...
+            string packageName, Guid projectGuid, EnvDTE.DTE dte, EnvDTE.Project dteProject,
+            IProgressTracker progressTracker, CancellationToken cancellationToken)
         {
-            this.AssertIsForeground();
             Contract.ThrowIfFalse(IsEnabled);
 
-            // TODO: consider putting this under a TWD so it would actually be cancellable.
-            var cancellationToken = CancellationToken.None;
+            // explicitly switch to BG thread to do the installation as nuget installer APIs are free threaded.
+            await TaskScheduler.Default;
```
Comment (sharwell) [inline src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs:410]: 💡 This is going to switch to the background thread and then _immediately_ switch back to the main thread to update the status bar. This line should be moved down to right before the `try`.

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs [hunk 10]
```diff
@@ -361,42 +391,56 @@ private void OnSourceProviderSourcesChanged(object sender, EventArgs e)
...
-                        var installedPackagesMap = await GetInstalledPackagesMapAsync(nugetService, projectGuid, cancellationToken).ConfigureAwait(true);
-                        if (!installedPackagesMap.TryGetValue(packageName, out var installedVersion))
-                            return false;
+                    cancellationToken.ThrowIfCancellationRequested();
 
-                        dte.StatusBar.Text = string.Format(ServicesVSResources.Uninstalling_0, packageName);
-                        _packageUninstaller.Value.UninstallPackage(dteProject, packageName, removeDependencies: true);
+                    // Once we start the installation, we can't cancel anymore.
+                    cancellationToken = default;
+                    _packageUninstaller.Value.UninstallPackage(dteProject, packageName, removeDependencies: true);
```
Comment (sharwell) [inline src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs:428]: ❗ This requires a preceding call to switch to the main thread until https://github.com/NuGet/Home/issues/11021 is complete. Adding explicit switches to the main thread for operations that are not yet documented will allow this refactoring to proceed without being blocked, and the subsequent change once the blocking issue is addressed will be minimal.

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs (hunk 5 above)
Comment (sharwell) [reply src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs:533]: > In this case, it is an MTA so you're fine.

@AArnott I wasn't able to find the documentation for this. Can you link it?

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs [hunk 11]
```diff
@@ -482,14 +526,15 @@ private void OnWorkspaceChanged(object sender, WorkspaceChangeEventArgs e)
         private async Task<T?> PerformNuGetProjectServiceWorkAsync<T>(
             Func<INuGetProjectService, CancellationToken, ValueTask<T?>> doWorkAsync, CancellationToken cancellationToken)
         {
+            // Make sure we are on the thread pool to avoid UI thread dependencies if external code uses ConfigureAwait(true).
+            // GetServiceAsync/GetProxyAsync and the cast below are all explicitly documented as being BG thread safe.
+            await TaskScheduler.Default;
```
Comment (sharwell) [inline src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs:531]: 📝 Would greatly prefer this context switch and comment be returned to its original location/form. The relocation doesn't improve performance of the code now under the umbrella of the switch, but it does move a required switch further away from the location where it is required.

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs [hunk 12]
```diff
@@ -291,65 +312,74 @@ private void OnSourceProviderSourcesChanged(object sender, EventArgs e)
...
-                        installedPackagesMap = await GetInstalledPackagesMapAsync(nugetService, projectGuid, cancellationToken).ConfigureAwait(true);
-                        var installedVersion = installedPackagesMap.TryGetValue(packageName, out var version) ? version : null;
-                        dte.StatusBar.Text = string.Format(ServicesVSResources.Installing_0_completed,
-                            GetStatusBarText(packageName, installedVersion));
-
-                        return true;
-                    }, cancellationToken);
-                });
+                        _packageInstaller.Value.InstallLatestPackage(
+                            source, dteProject, packageName, includePrerelease, ignoreDependencies: false);
```
Comment (sharwell) [inline src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs:357]: ❗ This requires a preceding call to switch to the main thread until NuGet/Home#11021 is complete (see other comment on the subject)

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs [hunk 13]
```diff
@@ -291,65 +312,74 @@ private void OnSourceProviderSourcesChanged(object sender, EventArgs e)
...
+                    }
+                    else
+                    {
+                        _packageInstaller.Value.InstallPackage(
+                            source, dteProject, packageName, version, ignoreDependencies: false);
```
Comment (sharwell) [inline src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs:362]: ❗ This requires a preceding call to switch to the main thread until NuGet/Home#11021 is complete (see other comment on the subject)

**Review by sharwell — CHANGES_REQUESTED** [review]
Overall the change is fine as a refactoring. Since we are blocked on NuGet/Home#11021 for the current form, I recommend the following:
1. Add explicit calls to `SwitchToMainThreadAsync` before the NuGet operations which are not yet documented
2. Retitle this pull request to indicate that it is a refactoring, since it will no longer be changing the threads used for installation/uninstallation of packages
3. File a follow-up issue to change the `SwitchToMainThreadAsync` calls over to `TaskScheduler.Default` once the blocking issue is addressed

This approach will expedite this pull request by side-stepping the blocking issue.

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs (hunk 12 above)
Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs:357]: this PR is being written with teh assumption that the doc changes will go in before this is done.

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs (hunk 9 above)
Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs:410]: seems reasonable.  but in that case, why do we need the switch at all?  the UpdateStatusBarAsync call has CA(false) on it.  So we'll be off hte UI thread then, right?

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs (hunk 10 above)
Comment (sharwell) [reply src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs:428]: ➡️ No change required here once NuGet/NuGet.Client#4209 merges.

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs (hunk 13 above)
Comment (sharwell) [reply src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs:362]: ➡️ No change required here once NuGet/NuGet.Client#4209 merges.

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs (hunk 12 above)
Comment (sharwell) [reply src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs:357]: ➡️ No change required here once NuGet/NuGet.Client#4209 merges.

**Comment by CyrusNajmabadi** [comment]
image

:'(

**Review by sharwell — APPROVED** [review]
Looks good now that https://github.com/NuGet/NuGet.Client/pull/4209 is moving forward

**Comment by CyrusNajmabadi** [comment]
@sharwell do you want me to wait on https://github.com/NuGet/NuGet.Client/pull/4209 being merged?  or are you ok knowing it's in progress?

**Comment by CyrusNajmabadi** [comment]
https://github.com/NuGet/NuGet.Client/pull/4209 merged in.  Merging this in now.
